src/coord/src/coord.rs

```diff
@@ -808,9 +808,21 @@ impl Coordinator {
                 appends.push((id, vec![], advance_to));
             }
         }
+
+        // Scratch vector holding all updates that happen at `t < advance_to`
+        let mut scratch = vec![];
         // Then drain all pending writes and prepare append commands
-        for (id, updates) in self.volatile_updates.drain() {
-            appends.push((id, updates, advance_to));
+        for (id, updates) in self.volatile_updates.iter_mut() {
+            // TODO(petrosagg): replace with `drain_filter` once it stabilizes
+            let mut cursor = 0;
+            while let Some(update) = updates.get(cursor) {
+                if update.timestamp < advance_to {
```

Comment:
My understanding is that these updates are coming from the `send_builtin_table_updates_at_offset` function, where we use some future timestamp to correctly retract data without possibility for forgetting it. This PR removes the second half of that, because now the data are split and backed by a volatile in memory data structure. We (coord) don't have plans on making that WAL-backed for now (it's hard and we don't need it for any upcoming milestone).

Reply:
These updates come from anything that write to tables, including what you said but also user `INSERT` statements, which must conform to the semantics of the `Append` operation. Is there another way to cut the batches that makes coord happy?

Reply:
I think that between the various needs here, `send_builtin_table_updates_at_offset` is on the losing end and should be changed to not do its magical timestamp retraction thing. If we commit this PR as is, I think any coord restart will always create some permanently wrong metrics.

Reply:
Wait, how do INSERTs generate timestamps that would trigger this?

Reply:
Does it make sense to tag in @jkosh44 here? There are various other "tables have the wrong data on restart" issues that he's looked at, where one conclusion was that it might be most ergonomic to support e.g. `truncate(table_id)`.

But, we should def see if we agree on the API and that/whether the calls are each meant to be durable (even if they are not currently so).

Reply:
My understanding is that the TimestampOracle does correct things such that `advance_local_inputs` would always commit all INSERT data, even with this PR. Is that not the case?

Reply:
No that is not the case, and it's related to my own misconception of what the `TimestampOracle` does. We currently call `advance_local_inputs` whenever the oracle returns `Some(ts)` from its `should_advance_to` call. This however will return the same timestamp that is currently used for writing, if we are in writing mode.

This means that we are only able to append to tables the INSERT data that has been written in strictly previous times then the current time of the oracle. If the oracle is in reading mode then you're correct that all INSERT data is appended to the tables. If the oracle is in writing mode however we must exclude the timestamp that is currently being written, and therefore there might be some pending INSERT data that has to wait until the next time we decide to read.

I looked into `send_builtin_table_updates_at_offset` and we were only calling it with an offset of zero so I removed it in this PR and did the work directly in `send_builtin_table_updates`